src/common/charconv.h

```diff
@@ -0,0 +1,78 @@
+/*!
+ * Copyright 2019 by XGBoost Contributors
+ */
+#ifndef XGBOOST_COMMON_CHARCONV_H_
+#define XGBOOST_COMMON_CHARCONV_H_
+
+#include <cstddef>
+#include <system_error>
+#include <iterator>
+#include <limits>
+
+#include "xgboost/logging.h"
+
+namespace xgboost {
+int32_t ToCharsFloatImpl(float f, char * const result);
+
+struct to_chars_result {  // NOLINT
+  char* ptr;
+  std::errc ec;
+};
+
+template <typename T>
+struct NumericLimits;
+
+template <> struct NumericLimits<float> {
+  // Unlike std::numeric_limit<float>::max_digits10, which represents the **minimum**
+  // length of base10 digits that are necessary to uniquely represent all distinct values.
+  // This value is used to represent the maximum length.  As sign bit occupies 1 character:
+  // sign + len(str(2^24)) + decimal point + `E` + sign + len(str(2^8)) + '\0'
+  static constexpr size_t kMaxDigit10Len = 16;
```

Comment:
So `std::numeric_limit<float>::max_digits10` is not sufficient? Do you have a reference to this?

Reply:
I don't have a reference, I calculated myself based on the algorithm.  Here is an example `1.9310392E-38`, `std::numeric_limit<float>::max_digits10` equals to `9`, which is not sufficient to print the value.